File: cpp/DailyTemperatures.cpp
#include <iostream> 
#include <stack>
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        stack<pair<int, int>> stack;
        vector<int> result(n);
        
        for (int i = 0; i < n; i++) {
            int currDay = i;
            int currTemp = temperatures[i];
            
            while (!stack.empty() && stack.top().second < currTemp) {
                int prevDay = stack.top().first;
                int prevTemp = stack.top().second;
                stack.pop();
                
                result[prevDay] = currDay - prevDay;
            }
            
            stack.push({currDay, currTemp});
        }
        
        return result;
    }
};

File: python/DailyTemperatures.py
from typing import List
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        resArr = [0] * len(temperatures)
        stack = []

        # iterate thru the list, add values to a stack until the higher temp is not met, then update res if higher one is found
        # we use a monotonicly decreasing stack, since its a very efficient way to solve the problem
        for i in range(len(temperatures)):
            while stack and stack[-1][0] < temperatures[i]:
                stackT, stackI = stack.pop()
                resArr[stackI] = i - stackI

            stack.append([temperatures[i], i])
        
        return resArr
            